internal/sidekick/rust/annotate.go

```diff
@@ -690,6 +701,119 @@ func (c *codec) addFeatureAnnotations(model *api.API, ann *modelAnnotations) {
+func (c *codec) findResourceNameCandidates(m *api.Method) []*resourceNameCandidateField {
+	var candidates []*resourceNameCandidateField
+
+	// Find top-level annotated fields
+	for _, field := range m.InputType.Fields {
+		if field.IsResourceReference && !field.Repeated && !field.Map && field.Typez == api.STRING_TYPE {
+			candidates = append(candidates, &resourceNameCandidateField{
+				FieldPath: []string{field.Name},
+				Field:     field,
+				IsNested:  false,
+				Accessor:  makeFieldAccessor(field, "req"),
+			})
+		}
+	}
+
+	// Find nested annotated fields (one level deep)
+	for _, field := range m.InputType.Fields {
+		if field.MessageType == nil || field.Repeated || field.Map {
+			continue
+		}
+		for _, nestedField := range field.MessageType.Fields {
+			if !nestedField.IsResourceReference || nestedField.Repeated || nestedField.Map || nestedField.Typez != api.STRING_TYPE {
+				continue
+			}
+			parentAccessor := makeFieldAccessor(field, "req")
+			nestedAccessor := makeFieldAccessor(nestedField, "s")
+
+			// Combine accessors: parent.and_then(|s| nested)
+			fullAccessor := fmt.Sprintf("%s.and_then(|s| %s)", parentAccessor, nestedAccessor)
+
+			candidates = append(candidates, &resourceNameCandidateField{
+				FieldPath: []string{field.Name, nestedField.Name},
+				Field:     nestedField,
+				IsNested:  true,
+				Accessor:  fullAccessor,
+			})
+		}
+	}
+	return candidates
+}
+
+func (c *codec) findResourceNameFields(m *api.Method) []*resourceNameCandidateField {
+	if m.InputType == nil {
+		return nil
+	}
+
+	candidates := c.findResourceNameCandidates(m)
+
+	if len(candidates) == 0 {
+		return nil
+	}
+
+	// Check for HTTP path presence
+	var httpParams map[string]bool
+	if m.PathInfo != nil && m.PathInfo.Codec != nil {
+		if pia, ok := m.PathInfo.Codec.(*pathInfoAnnotation); ok {
+			httpParams = make(map[string]bool)
+			for _, p := range pia.UniqueParameters {
+				httpParams[p.FieldName] = true
+			}
+		}
+	}
+
+	isInPath := func(c *resourceNameCandidateField) bool {
+		if httpParams == nil {
+			return false
+		}
+		var snakeParts []string
+		for _, p := range c.FieldPath {
+			snakeParts = append(snakeParts, toSnake(p))
+		}
+		fieldName := strings.Join(snakeParts, ".")
+		return httpParams[fieldName]
+	}
+
+	// Sort candidates by priority:
+	// 1. Top-level fields (IsNested == false)
+	// 2. Fields in HTTP path (isInPath == true)
+	// 3. Proto definition order (stable sort)
+	sort.SliceStable(candidates, func(i, j int) bool {
+		// 1. Top-level before Nested
+		if candidates[i].IsNested != candidates[j].IsNested {
+			return !candidates[i].IsNested // false (top) < true (nested)
+		}
+		// 2. In-Path before Not-In-Path
+		inPathI := isInPath(candidates[i])
+		inPathJ := isInPath(candidates[j])
+		if inPathI != inPathJ {
+			return inPathI // true (in path) < false (not in path) -> we want true first
+		}
+		// 3. Stable sort preserves proto order
+		return i < j
+	})
```

Comment:
The sorting function looks complicated enough to have a name and (hopefully smaller) unit tests?